leetcode/medium/498._Diagonal_Traverse.py

```diff
@@ -0,0 +1,67 @@
+# Link to problem: https://leetcode.com/problems/diagonal-traverse/description/
+# Loom explaining solution you wrote (5m): TODO
+# Solution:
+# Description
+# Key ideas
+# Reason through the boundary conditions by drawing a bigger square with arrows crossing through them
+# There’s nothing difficult but one needs to be extra careful
+# Difficulty: Medium
+# Code
+
+class Solution:
+    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
+        def next_up_cord(cord):
+            x, y = cord
+            return [x + 1, y - 1]
+
+        def next_down_cord(cord):
+            x, y = cord
+            return [x - 1, y + 1]
+
+        def out_of_bound(cord):
+            x, y = cord
+            if x < 0 or y < 0:
+                return True
+            if y >= len(mat):
+                return True
+            if x >= len(mat[0]):
+                return True
+            return False
+
+        def at(cord):
+            x, y = tuple(cord)
+            return mat[y][x]
+
+        up = True
+        res = []
+        cord = [0, 0]
+        while cord != (len(mat[0]), len(mat)):
```

Comment:
Another way to think about this problem is to treat it like spiral matrix.

1. Think of each cell as a node.
2. You have the following directions. `right, down left, down, up right` you can initialize an array with a function that determines the next direction based on the current position and direction.
3. You pseudocode can follow the following format.

1. while current is not the target.
2. travel in the current direction until I hit a wall
3. Once I hit a wall, change directions accordingly (use a helper function)
4. repeat step 2 until we get to the target.

Reply:
Let's implement this approach as well as your current one

Reply:
Isn't this the same as my approach? The only difference I see here is
1. instead of checking if we hit the wall, I check if I've gone over the wall (out of bound)